### DISCUSSION TITLE
xwax Implementation

### DISCUSSION THREAD

**Discussion by riczikovacs** [SynthstromAudible/DelugeFirmware#277] [Ideas]
**Idea**
I am thinking about trying to port over the timecode vinyl engine from xwax to the deluge to control stuff. [https://github.com/xwax/xwax](https://github.com/xwax/xwax)
The initial idea would be to control the note on/off, start position, play direction and pitch/speed of a audio track or possibly sample voice in real time. Here are some considerations I have been pondering, but I am extremally curious what everyone thinks.

**Legal**
xwax uses GNU General Public License V3. I may be totally mixed up but I recall from the livestreams that using code from this kind of license is ok for the community branch. However a separate license would need to be obtained if the feature were to be merged into the official branch. If anyone knows better about the license PLEASE educate me!

**Usage**
Not really sure if this would be interesting to others, but I would love to be able to use the Deluge as a standalone timecode vinyl player. As far as I know there really is not such a thing on the market at the moment outside using a raspberry pi or something. The timecode signal would be read through the line or mic in. Control could be assigned to a voice via the learn button, other settings such as Serrato vs Traktor etc. can live in a global submenu. Realtime control of a single voice only at first, see technical section. Future ideas are endless however, control the playback of a track Euclidean style? Send that vinyl controlled track to an external synth over midi?? Control the tempo and playhead of an entire song or arranger mode!!?? Utilize some of the stutter buffer and re pitch code for real time vinyl emulation of same!!!!!?????? 

**Technical**
No Recording?? The proposed control will have to access a few things that are not normally midi controllable, namely start position and play direction. I think control of the start position may have been added in the community branch already... really not sure about play direction entirely. Therefore to begin with I am shooting for Realtime control only, as recording would require adding recording support for those locked parameters... maybe out of scope for the moment.

Here is a really cool rundown of how timecode vinyl works [https://mixxx.discourse.group/t/how-does-timecode-vinyl-actually-work-pt-2/23844](https://mixxx.discourse.group/t/how-does-timecode-vinyl-actually-work-pt-2/23844)

The existing analog input buffer could be read into the xwax decode algorithms which should be pretty fast. Looks like the largest lfsr lookup table xwax uses is about 2 megabytes. Not sure about other ram requirements but it doesn't seem out of the question... Some feasibility testing will definitely have to be done. From there I think it is a matter of mapping the control values and let the scritchy scratchy madness begin!

**Feedback**
I am super interested in what anyone thinks about all aspects of this idea, even if it is to shoot it down I really appreciate the feedback! This would be a long term project with many steps, so for now please consider it just a funky idea to be floated around.

**Comment by sapphire-arches** [comment]
If I understand correctly, the idea would be to take a timecode record in on the line-in and use that to control playback of an audio sample/sequence? As you point out it looks like xwax is licensed compatibly. However, I think code and RAM size will be a real challenge. The RZ/A1L used in the Deluge has only 3MB of RAM, and about half of that is already consumed by existing Deluge code. Any audio code pretty much has to execute out of RAM and we only have 3MB of flash space total to use. Perhaps the LFSR identification could be done with less memory (but likely higher CPU usage), and the timestretching/pitch shifting could be done using some of the existing DSP code to further save on memory? This would likely entail an entirely from-scratch implementation of the timecode logic and DSP however.

A feature like this does seem cool and there are people around who could help guide you, but it seems like it will be quite the challenge.

An alternative approach would be throwing out the all but the lowest levels of the existing firmware and using the Deluge as a blank slate hardware platform that does nothing but handle vinyl timecode -- this might be more feasible as you could make fuller use of the hardware than you could while working within the exiting firmware.

**Reply by riczikovacs** [reply]
Thank you so much for the response! I was thinking of the 64 mb of sdram when looking at this. It sounds like that would not be fast enough to search the lfsr table? Maybe one of the open DMAs would help?

I also find your idea of a complete firmware version focused only on this feature pretty interesting. I wonder if it would still perform well enough to keep a few tracks playing and run some effects, thereby making a useable "DJ Deluge" firmware version. 

Again, thank you so much for your effort and insight looking at this! Sounds like I have some research to do yet...

**Reply by sapphire-arches** [reply]
The sdram could fit the LFSR reference table yeah, and it should be fast enough. My concern with the size is what would happen if you just built the xwax code and tried to load it =). The table could be loaded off of the sdcard though, as the internal flash would be too small to contain it. I don't know how much a DMA channel would help as the DMA engines just move bytes -- they can't be used for any sort of logic (though DMA is used to accelerate loading data off of the SDCard as well as audio input). I think there's enough juice in the CPU to run timecode decode + retime on 1 audio channel, and maybe a few low-complexity synth voices. At least, I can't immediately think of a reason some sort of functionality like this is not possible. Getting it all to fit in terms of both code size and performance headroom will be a real challenge, but it feels at least possible.

**Reply by Quadristi-Riczi** [reply]
Thanks again for the feedback!  Just received my debugger parts and some serato control vinyls :) Ill probably start by just reading out decoded position, but controlling an audio track does seem like a logical next step.

**Comment by worker-gray** [comment]
Is anything developing on this? This sounds like something I've been requesting for years!

**Reply by riczikovacs** [reply]
Thanks for the interest!! I have all the hardware to start development but have not gone through setting up the toolchain for vscode yet... I have a similar work project that has been sucking that particular creative juice. I have learnt a lot that will help with implementing this feature though.

**Reply by worker-gray** [reply]
that's great news! I started a thread on 'scratching' samples with gold knobs. It has some 'ideas' on implementation.
https://github.com/SynthstromAudible/DelugeFirmware/discussions/1184

**Comment by worker-gray** [comment]
has this project been abandoned?
